changefeedccl: reduce rebalancing memory usage from O(ranges) to O(spans)

jayshrivastava opened the pull request:
### sql: count ranges per partition in PartitionSpans

This change updates span partitioning to count ranges while making
partitions. This allows callers to rebalance partitions based on
range counts without having to iterate over the spans to count
ranges.

Release note: None
Epic: None

### changefeedccl: reduce rebalancing memory usage from O(ranges) to O(spans) #115375

Previously, the `rebalanceSpanPartitions` would use O(ranges) memory. This change
rewrites it to use range iterators, reducing the memory usage to O(spans).

This change also adds a randomized test to assert that all spans are accounted for after
rebalancing. It also adds one more unit test.

Informs: https://github.com/cockroachdb/cockroach/issues/113898
Epic: None

### changefeedccl: add rebalancing checks

This change adds extra test coverage for partition rebalancing in
changefeeds. It adds checks which are performed after rebalancing
to assert that the output list of spans covers exactly the same keys
as the input list of spans. These checks are expensive so they only
run if the environment variable `COCKROACH_CHANGEFEED_TESTING_REBALANCING_CHECKS`
is true. This variable is true in cdc roachtests and unit tests.

Release note: None
Epic: None

jayshrivastava committed:
changefeedccl: reduce rebalancing memory usage from O(ranges) to O(spans)

Previously, the `rebalanceSpanPartitions` would use O(ranges) memory. This change
rewrites it to use range iterators, reducing the memory usage to O(spans).

This change also adds a randomized test to assert that all spans are accounted for after
rebalancing. It also adds one more unit test.

Informs: #113898
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/115375)
<!-- Reviewable:end -->

jayshrivastava commented:
Note that the first commit can be ignored as it's being reviewed separately in https://github.com/cockroachdb/cockroach/pull/115166

srosenberg reviewed (commented):
> Previously, the `rebalanceSpanPartitions` would use O(ranges) memory. This change  
rewrites it to use range iterators, reducing the memory usage to O(spans).

I am looking at the previous version, using `AllRangeSpans`. It also uses a range iterator. I guess I am not seeing where `O(ranges)` is being allocated.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-:-NkcLAGCD0Llu-Q2d7U1:b-87rb4v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @herkolategan, @miretskiy, and @renatolabs)



<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
If I'm reading the code correctly, `getRangesForSpans` returns a list of ranges. Then we save this list of ranges in `p[i].Spans` for each partition. So, we end up with all the ranges in memory stored as 1 span per range. This is the code:
```
numRanges := 0
	for i := range p {
		spans, err := r.getRangesForSpans(ctx, p[i].Spans)
		if err != nil {
			return nil, err
		}
		p[i].Spans = spans
		numRanges += len(spans)
	}
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-:-Nl4qsun-L4pL4sNrnZY:b-lcof6c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @herkolategan, @miretskiy, and @renatolabs)



<!-- Sent from Reviewable.io -->

srosenberg commented:
> If I'm reading the code correctly, `getRangesForSpans` returns a list of ranges. 

It returns a list of ranges _confined_ to the given spans, right?

> Then we save this list of ranges in `p[i].Spans` for each partition. So, we end up with all the ranges in memory stored as 1 span per range. 

True, we use `spans` as an intermediate buffer, which your PR removes. That's a great improvement! I was only confused by the commit message which says the improvement is from `O(ranges)` to `O(spans)`. I don't see an _asymptotic_ improvement in this PR; i.e., an improvement is in reducing the constant multiplier of the same asymptotic complexity.

jayshrivastava reviewed (commented):
I think you're right. It's hard to say because we cant express the number of ranges as a function of spans. Regardless I think O(ranges) -> O(spans) is a simple shorthand which gets the point across. It's nicer to have one span in memory than 250k ranges :)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-:-Nl53CjGB-TcKb_SlB3Q:b-bysfq9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @herkolategan, @miretskiy, and @renatolabs)



<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
You're definitely right. Ranges have a bounded size. If a table has a certain size in bytes, the number of ranges is a constant multiple of that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-:-NlA6qpz-xqhing9-X1_:b-liacq0)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @herkolategan, @miretskiy, and @renatolabs)



<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/ccl/changefeedccl/changefeed_dist.go:
```diff
@@ -545,80 +547,129 @@ var rebalanceThreshold = settings.RegisterFloatSetting(
+				return nil, err
+			}
+			// Iterate over the ranges in the current span.
+			for ri.Seek(ctx, rSpan.Key, kvcoord.Ascending); !terminate() && needMore(); ri.Next(ctx) {
```
i wonder if there is a way to do it in sql.PartitionSpans...
Going to think a bit more about this... but for now, flushing out comments so far.

jayshrivastava committed:
sql: count ranges per partition in `PartitionSpans`

This change updates span partitioning to count ranges while making
partitions. This allows callers to rebalance partitions based on
range counts without having to iterate over the spans to count
ranges.

Release note: None
Epic: None

jayshrivastava reviewed (commented):
I made is so that we don't need to count ranges in `rebalancePartitions`. Now, distsql gives us range counts along with the partitions. @miretskiy Let me know what you think!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-:-Nm2ndOvBOJ6rSojAI2R:b-jbx29e)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @herkolategan, and @renatolabs)



<!-- Sent from Reviewable.io -->

jayshrivastava committed:
changefeedccl: add rebalancing checks

This change adds extra test coverage for partition rebalancing in
changefeeds. It adds checks which are performed after rebalancing
to assert that the output list of spans covers exactly the same keys
as the input list of spans. These checks are expensive so they only
run if the environment variable `COCKROACH_CHANGEFEED_TESTING_REBALANCING_CHECKS`
is true. This variable is true in cdc roachtests and unit tests.

Release note: None
Epic: None

rharding6373 reviewed (commented):
Nice! Is there a microbenchmark that demonstrates these improvements?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-:-NssCRG0ErYorDbCmm58:b-4uhs2a)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @herkolategan, @jayshrivastava, @miretskiy, and @renatolabs)

___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 563 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsnVJ4b2GcTLokN4bVe:-NsnVJ4cC52CG55lVqA9:b-a8wi1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/af156d2ce68f7536e15bfb90a1a724fcc0de3618/pkg/ccl/changefeedccl/changefeed_dist.go#L563)):*
> ```Go
> 	// and mutated during rebalancing.
> 	numRanges int
> 	g         roachpb.SpanGroup
> ```

nit: use more descriptive names for struct members like `group` and `part`
___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 587 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NspCaEI2LDbpjDUOX--:-NspCaEI2LDbpjDUOX-0:bj2y8uf) ([raw file](https://github.com/cockroachdb/cockroach/blob/af156d2ce68f7536e15bfb90a1a724fcc0de3618/pkg/ccl/changefeedccl/changefeed_dist.go#L587)):*
> ```Go
> 		// We cannot rebalance if we're missing range information.
> 		if !ok {
> 			log.Infof(ctx, "skipping rebalance due to missing range info")
> ```

This seems worth a warning.
___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 600 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NssAVGo3AoKDdZvArIx:-NssAVGo3AoKDdZvArIy:bnkwedv) ([raw file](https://github.com/cockroachdb/cockroach/blob/af156d2ce68f7536e15bfb90a1a724fcc0de3618/pkg/ccl/changefeedccl/changefeed_dist.go#L600)):*
> ```Go
> 	})
> 
> 	targetRanges := int(math.Ceil((1 + sensitivity) * float64(totalRanges) / float64(len(partitions))))
> ```

For my education, how do we normally determine sensitivity? I noticed that in the test sensitivity is 0 and there's a TODO to get rid of it. Does that affect only the sensitivity in the test? Do we test that the algorithm works with sensitivity in another test?
___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 667 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsnP9IU6dZKBdyQdXg6:-NsnP9IU6dZKBdyQdXg7:bqglitl) ([raw file](https://github.com/cockroachdb/cockroach/blob/af156d2ce68f7536e15bfb90a1a724fcc0de3618/pkg/ccl/changefeedccl/changefeed_dist.go#L667)):*
> ```Go
> 	for _, b := range builders {
> 		partitions[b.pIdx] = sql.MakeSpanPartition(
> 			b.p.SQLInstanceID, b.g.Slice(), true, b.numRanges)
> ```

nit: add comment with param name to bool arg for all these calls.
___
*[`pkg/sql/distsql_physical_planner.go` line 1222 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NssBkj_9R_j5dSCNuFI:-NssBkj_9R_j5dSCNuFJ:bykqlui) ([raw file](https://github.com/cockroachdb/cockroach/blob/615a574860e4301d0ceaf94536f2f9ef1b226309/pkg/sql/distsql_physical_planner.go#L1222)):*
> ```Go
> // MakeSpanPartition constructs a SpanPartition.
> func MakeSpanPartition(
> 	instanceID base.SQLInstanceID, spans roachpb.Spans, haveRangeInfo bool, numRanges int,
> ```

In all calls to `MakeSpanPartition`, `haveRangeInfo` is true. I would get rid of the bool param and just put true in the struct below.
___
*[`pkg/ccl/changefeedccl/changefeed_dist_test.go` line 136 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NspH7a0CMhfiQaEn0rY:-NspH7a0CMhfiQaEn0rZ:b-9857ah) ([raw file](https://github.com/cockroachdb/cockroach/blob/af156d2ce68f7536e15bfb90a1a724fcc0de3618/pkg/ccl/changefeedccl/changefeed_dist_test.go#L136)):*
> ```Go
> 				[]roachpb.Span{mkRange('z' - i)},
> 				true,
> 				1,
> ```

nit: the constant arg could also use a param comment
___
*[`pkg/ccl/changefeedccl/changefeed_dist_test.go` line 268 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NspGxMh01RTWaVO08wt:-NspGxMiAhj5ENtZr3Sy:bds3i29) ([raw file](https://github.com/cockroachdb/cockroach/blob/af156d2ce68f7536e15bfb90a1a724fcc0de3618/pkg/ccl/changefeedccl/changefeed_dist_test.go#L268)):*
> ```Go
> 	// Create a random input and assert that the output has the same
> 	// spans as the input.
> 	t.Run("random", func(t *testing.T) {
> ```

Nice test!
___
*[`pkg/ccl/changefeedccl/changefeed_dist_test.go` line 322 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsokfXY1s03txqSSTYH:-NsokfXY1s03txqSSTYI:b-e2no1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/af156d2ce68f7536e15bfb90a1a724fcc0de3618/pkg/ccl/changefeedccl/changefeed_dist_test.go#L322)):*
> ```Go
> 		g2 := copySpans(output)
> 		require.True(t, g1.Encloses(g2.Slice()...))
> 		require.True(t, g2.Encloses(g1.Slice()...))
> ```

Is there another check here we can make around how many spans per partition there are? Assert some max threshold is not exceeded, maybe?


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
Thanks for giving this a look!

Unfortunately, I don't have a microbenchmark. That being said, I'm pretty confident in this. The old code would load all the ranges into memory. Now we load spans into memory.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-:-Nst1HUE8-tsYvFDemUH:b71dp2v)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @herkolategan, @miretskiy, @renatolabs, and @rharding6373)

___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 600 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/115375#gh-1421024540:-Nssu65gDR72KU4mth7L:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3e74a45f24fc24c69f532fe7f6dda752eed0a1d/pkg/ccl/changefeedccl/changefeed_dist.go#L600)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

it's a bit sad that we have to have all this code just to count the number of ranges.
I think we can easily extend `sql.SpanPartition` struct to have num ranges counters which we can update as we're building span partitions.

This will remove the need to create 2 iterators.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 612 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/115375#gh-1421006815:-NssuAXX8tojN1kPYgv3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3e74a45f24fc24c69f532fe7f6dda752eed0a1d/pkg/ccl/changefeedccl/changefeed_dist.go#L612)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

terminate is so.... fatal... Maybe `canRebalance`?
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 639 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/115375#gh-1421049862:-Nsszl0bEAde3ZqK_gcO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3e74a45f24fc24c69f532fe7f6dda752eed0a1d/pkg/ccl/changefeedccl/changefeed_dist.go#L639)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

i wonder if there is a way to do it in sql.PartitionSpans...
Going to think a bit more about this... but for now, flushing out comments so far.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 587 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NspCaEI2LDbpjDUOX--:-Nssu3nH2P20mKNNoCOu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af156d2ce68f7536e15bfb90a1a724fcc0de3618/pkg/ccl/changefeedccl/changefeed_dist.go#L587)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

This seems worth a warning.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 600 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NssAVGo3AoKDdZvArIx:-Nst-OfCDuTEe1hIqjhW:birq1zy) ([raw file](https://github.com/cockroachdb/cockroach/blob/af156d2ce68f7536e15bfb90a1a724fcc0de3618/pkg/ccl/changefeedccl/changefeed_dist.go#L600)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

For my education, how do we normally determine sensitivity? I noticed that in the test sensitivity is 0 and there's a TODO to get rid of it. Does that affect only the sensitivity in the test? Do we test that the algorithm works with sensitivity in another test?
</blockquote></details>

It's changed by the setting `changefeed.balance_range_distribution.sensitivity`. Some of the other, larger tests in changefeed_dist_test.go use the default sensitivity of 5%.

I filed https://github.com/cockroachdb/cockroach/issues/120427. I think we should actually change it to some other sort of sensitivity (ex. if the number of ranges is below X, then don't rebalance or rebalance in a different way). Something like this would be more useful, but we need to do some thinking/design first.
___
*[`pkg/sql/distsql_physical_planner.go` line 1222 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NssBkj_9R_j5dSCNuFI:-NssycJ91p_gtD99v_N-:bow65to) ([raw file](https://github.com/cockroachdb/cockroach/blob/615a574860e4301d0ceaf94536f2f9ef1b226309/pkg/sql/distsql_physical_planner.go#L1222)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

In all calls to `MakeSpanPartition`, `haveRangeInfo` is true. I would get rid of the bool param and just put true in the struct below.
</blockquote></details>

Done. I changed the name as well.
___
*[`pkg/ccl/changefeedccl/changefeed_dist_test.go` line 268 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NspGxMh01RTWaVO08wt:-Nst-NSy7GF_Tt0z2s8h:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/af156d2ce68f7536e15bfb90a1a724fcc0de3618/pkg/ccl/changefeedccl/changefeed_dist_test.go#L268)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Nice test!
</blockquote></details>

Thanks!
___
*[`pkg/ccl/changefeedccl/changefeed_dist_test.go` line 322 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsokfXY1s03txqSSTYH:-Nsszr48-iaqKbucCKEb:bqzasvw) ([raw file](https://github.com/cockroachdb/cockroach/blob/af156d2ce68f7536e15bfb90a1a724fcc0de3618/pkg/ccl/changefeedccl/changefeed_dist_test.go#L322)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Is there another check here we can make around how many spans per partition there are? Assert some max threshold is not exceeded, maybe?
</blockquote></details>

Encloses also checks the spans contained not just the count. I believe that I asked in slack how to check two sets of spans for equivalency and someone mentioned that doing the double enclose is the way to do it.


<!-- Sent from Reviewable.io -->

rharding6373 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115375)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @herkolategan, @jayshrivastava, @miretskiy, and @renatolabs)

___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 600 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NssAVGo3AoKDdZvArIx:-Nst4Ot80LHDo-IuW45N:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/af156d2ce68f7536e15bfb90a1a724fcc0de3618/pkg/ccl/changefeedccl/changefeed_dist.go#L600)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

It's changed by the setting `changefeed.balance_range_distribution.sensitivity`. Some of the other, larger tests in changefeed_dist_test.go use the default sensitivity of 5%.

I filed https://github.com/cockroachdb/cockroach/issues/120427. I think we should actually change it to some other sort of sensitivity (ex. if the number of ranges is below X, then don't rebalance or rebalance in a different way). Something like this would be more useful, but we need to do some thinking/design first.
</blockquote></details>

Thanks!
___
*[`pkg/ccl/changefeedccl/changefeed_dist_test.go` line 322 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsokfXY1s03txqSSTYH:-Nst4a10CJVw38NtqRAX:bv40c9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/af156d2ce68f7536e15bfb90a1a724fcc0de3618/pkg/ccl/changefeedccl/changefeed_dist_test.go#L322)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

Encloses also checks the spans contained not just the count. I believe that I asked in slack how to check two sets of spans for equivalency and someone mentioned that doing the double enclose is the way to do it.
</blockquote></details>

To clarify, my question is whether there is a way to verify that the algorithm rebalanced so that all partitions have an equal number of spans. Encloses just shows that the same set of spans exist before and after, right?


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115375)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @herkolategan, @miretskiy, @renatolabs, and @rharding6373)

___
*[`pkg/ccl/changefeedccl/changefeed_dist_test.go` line 322 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsokfXY1s03txqSSTYH:-NstcGSh0coMEiUsqIdi:b-pldpse) ([raw file](https://github.com/cockroachdb/cockroach/blob/af156d2ce68f7536e15bfb90a1a724fcc0de3618/pkg/ccl/changefeedccl/changefeed_dist_test.go#L322)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

To clarify, my question is whether there is a way to verify that the algorithm rebalanced so that all partitions have an equal number of spans. Encloses just shows that the same set of spans exist before and after, right?
</blockquote></details>

Good point! I updated the code to check the range counts.


<!-- Sent from Reviewable.io -->

rharding6373 reviewed (approved):
LGTM Nice work!

andyyang890 reviewed (commented):
> Reviewed 1 of 11 files at r5, 1 of 5 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115375)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @herkolategan, @jayshrivastava, @miretskiy, @renatolabs, and @rharding6373)

___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 602 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsvX137ClETbeRosSp3:-NsvX137ClETbeRosSp4:b-owwsni) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c8b3014c0c918aa490a587dc7349cda6d51eb46/pkg/ccl/changefeedccl/changefeed_dist.go#L602)):*
> ```Go
> 	}
> 
> 	//Create partition builder structs for the partitions array above.
> ```

nit: missing space
___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 611 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsvWu97CnX_SIGzivzb:-NsvWu97CnX_SIGzivzc:bih7qya) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c8b3014c0c918aa490a587dc7349cda6d51eb46/pkg/ccl/changefeedccl/changefeed_dist.go#L611)):*
> ```Go
> 		// We cannot rebalance if we're missing range information.
> 		if !ok {
> 			log.Warning(ctx, "skipping rebalance due to missing range info")
> ```

Should we maybe fallback on the old code instead of not rebalancing if range info is missing? I worry that hypothetically if a bug caused the range info not to be populated, we would lose the ability to rebalance at all.
___
*[`pkg/sql/distsql_physical_planner.go` line 1209 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsvTweAA_OFYkYW9QZB:-NsvTweAA_OFYkYW9QZC:b-ph5z9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/fac74ec02cfa133a9b7cd62525e3c4b75e3c636c/pkg/sql/distsql_physical_planner.go#L1209)):*
> ```Go
> 
> 	haveRangeInfo bool
> 	numRanges     int
> ```

Thoughts on making this field a `*int` so you don't need `haveRangeInfo`? If you do decide to go down that route, you could also have a `numRangesAdd` receiver to make the pointer dereferencing needed for adding a little simpler.
___
*[`pkg/sql/distsql_physical_planner_test.go` line 1237 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsvVSc3DVH7buI2htPn:-NsvVSc3DVH7buI2htPo:b-50tt2f) ([raw file](https://github.com/cockroachdb/cockroach/blob/fac74ec02cfa133a9b7cd62525e3c4b75e3c636c/pkg/sql/distsql_physical_planner_test.go#L1237)):*
> ```Go
> 				rangeCount += n
> 			}
> 			require.Equal(t, countRanges(partitions), rangeCount)
> ```

Should this test maybe assert equality for each partition instead of just the total?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
LGTM.

> Reviewed 1 of 11 files at r5, 1 of 5 files at r8, 4 of 4 files at r11, 3 of 3 files at r12, 2 of 2 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-:-NsyWOUf7wFB5RLwJfG5:b-85i63v)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @herkolategan, @jayshrivastava, @miretskiy, @renatolabs, and @rharding6373)

___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 682 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsyUt9KAH2I5lQV374q:-NsyUt9KAH2I5lQV374r:bqgnpsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec358ec9db441a75f9e1922ace36d605273a0022/pkg/ccl/changefeedccl/changefeed_dist.go#L682)):*
> ```Go
> 	// Overwrite the original partitions slice with the balanced partitions.
> 	for _, b := range builders {
> 		partitions[b.pIdx] = sql.MakeSpanPartitionWithRangeCount(
> ```

For my own edification - IIUC we don't really need to store the updated range count information, right? In other words, we used this information during the rebalancing loop and we won't use it in the future, correct?
___
*[`pkg/sql/distsql_physical_planner.go` line 1371 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsyRQSECLx_aO5FNSH3:-NsyRQSECLx_aO5FNSH4:b-s5rq2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/fac74ec02cfa133a9b7cd62525e3c4b75e3c636c/pkg/sql/distsql_physical_planner.go#L1371)):*
> ```Go
> 	getSQLInstanceIDForKVNodeID func(roachpb.NodeID) (base.SQLInstanceID, SpanPartitionReason),
> 	ignoreMisplannedRanges *bool,
> ) (_ []SpanPartition, lastPartitionIdx int, err error) {
> ```

nit: why give the error return argument a name?
___
*[`pkg/ccl/changefeedccl/changefeed_dist_test.go` line 125 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsyVu7h23VybUN1WzHt:-NsyVu7h23VybUN1WzHu:b7rrf3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec358ec9db441a75f9e1922ace36d605273a0022/pkg/ccl/changefeedccl/changefeed_dist_test.go#L125)):*
> ```Go
> // TestPartitionSpans unit tests the rebalanceSpanPartitions function.
> func TestPartitionSpans(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> ```

nit: now that the test logs more stuff consider adding `defer log.Scope(t).Close(t)`.


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115375)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @DrewKimball, @herkolategan, @miretskiy, @renatolabs, @rharding6373, and @yuzefovich)

___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 682 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsyUt9KAH2I5lQV374q:-NtH4LnkEFMZ2YDxREcu:bb77ufg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec358ec9db441a75f9e1922ace36d605273a0022/pkg/ccl/changefeedccl/changefeed_dist.go#L682)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

For my own edification - IIUC we don't really need to store the updated range count information, right? In other words, we used this information during the rebalancing loop and we won't use it in the future, correct?
</blockquote></details>

Yes.
___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 611 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsvWu97CnX_SIGzivzb:-NtH3V7TDAUzkZX16ZSJ:bi0qkam) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c8b3014c0c918aa490a587dc7349cda6d51eb46/pkg/ccl/changefeedccl/changefeed_dist.go#L611)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

Should we maybe fallback on the old code instead of not rebalancing if range info is missing? I worry that hypothetically if a bug caused the range info not to be populated, we would lose the ability to rebalance at all.
</blockquote></details>

I think this would only happen if someone made a code change, but tests would fail if such a change was made. I don't think it's possible for the info to be missing for any other reason. Distsql does it while updating the `Spans` in the partition: https://github.com/cockroachdb/cockroach/blob/fac74ec02cfa133a9b7cd62525e3c4b75e3c636c/pkg/sql/distsql_physical_planner.go#L1363. There's no way for spans to be there without range info. I left the warning there just in case but it really shouldn't happen. 
___
*[`pkg/sql/distsql_physical_planner.go` line 1209 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsvTweAA_OFYkYW9QZB:-NtH4RN73V4oKRRtAAaC:b-rt6u6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/fac74ec02cfa133a9b7cd62525e3c4b75e3c636c/pkg/sql/distsql_physical_planner.go#L1209)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

Thoughts on making this field a `*int` so you don't need `haveRangeInfo`? If you do decide to go down that route, you could also have a `numRangesAdd` receiver to make the pointer dereferencing needed for adding a little simpler.
</blockquote></details>

I don't really want to make this change for an int. Feels wrong to put the int field on the heap unnecessarily. I know that avro/json does it but I think that's because they really need to.
___
*[`pkg/sql/distsql_physical_planner_test.go` line 1237 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/115375#-NsvVSc3DVH7buI2htPn:-NtH5xzo4e7KzEHAIh9O:bn7pclr) ([raw file](https://github.com/cockroachdb/cockroach/blob/fac74ec02cfa133a9b7cd62525e3c4b75e3c636c/pkg/sql/distsql_physical_planner_test.go#L1237)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

Should this test maybe assert equality for each partition instead of just the total?
</blockquote></details>

It does right here https://github.com/cockroachdb/cockroach/blob/fac74ec02cfa133a9b7cd62525e3c4b75e3c636c/pkg/sql/distsql_physical_planner_test.go#L1239-L1267


<!-- Sent from Reviewable.io -->

jayshrivastava commented:
bors r+
